cranelift-wasm/src/state.rs

```diff
@@ -24,27 +24,41 @@ pub enum ControlStackFrame {
     If {
         destination: Ebb,
         branch_inst: Inst,
+        params: Vec<Value>,
```

Comment:
I'm concerned about the overhead of doing a dynamic allocation for each `If` and `Else` opcode that has parameters. Do you think it would be awkward to do something like this? On entry to an `If` with parameters, push a second copy of the parameter values onto the stack. Then, the `If` can use the second copy, and if there's an `Else`, it can use the first copy, and if not, we can just pop and discard the second copy.

Reply:
I can look into this. I'm a little concerned that it might be a little bit subtle / hard to follow for readers -- is translation usually CPU bottlenecked or IO/network bottlenecked?

Reply:
Ok this was actually pretty straightforward! Done.

Reply:
FWIW, cranelift performance in Firefox is currently limited in part by the absolutely massive volume of small-vector allocations in cranelift, and it would be nice not to make this worse.  Partly all the malloc/free behavior is overhead in itself, especially with poisioning-on-free; partly jemalloc makes it much worse by limiting parallel speedup due to its architecture.  cc @julian-seward1 who's been looking into this.  Also see [this bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1578352).

Reply:
I understand @sunfishcode's concern here, but I'd suggest to leave this as-is for the time being.  We can profile it later with dhat and if it does significantly contribute to malloc traffic then we can maybe change it to a SmallVec.  That has been effective in the past, per the bug that @lars-t-hansen references, but using SmallVec isn't free, in the sense it's necessary to establish an optimal-ish inline size, and also it adds a conditional branch to all accesses.  So let's wait for the data.

Reply:
Well, the double push of params to avoid the extra vec is already implemented, and I think it ended up being simpler than I had originally feared, so best of both worlds?